waku/v2/protocol/waku_store/waku_store.nim

```diff
@@ -457,9 +457,42 @@ proc query*(w: WakuStore, query: HistoryQuery, handler: QueryHandlerFunc) {.asyn
     return
 
   waku_store_messages.set(response.value.response.messages.len.int64, labelValues = ["retrieved"])
-
   handler(response.value.response)
 
+  
+proc findLastSeen*(list: seq[IndexedWakuMessage]): float = 
+  var lastSeenTime = float64(0)
+  for iwmsg in list.items : 
+    lastSeenTime = if iwmsg.msg.timestamp>lastSeenTime: iwmsg.msg.timestamp else: lastSeenTime 
```

Comment:
Not sure, but any way we could reconsider how we order messages in `ws.messages`?
We already compute an [epoch-based index for each message](https://github.com/status-im/nim-waku/blob/master/waku/v2/protocol/waku_store/waku_store.nim#L40). Elsewhere we [sort the messages](https://github.com/status-im/nim-waku/blob/master/waku/v2/protocol/waku_store/waku_store.nim#L256) in response to a `HistoryQuery` on index (which will have the effect of sorting them by time, due to the epoch indices). Here we cycle through all messages to find the youngest timestamp. All of these can be computationally quite expensive. Wouldn't it be easier to just maintain a time-based sorting order on `ws.messages`? In that case `findLastSeen` would just be the timestamp of the last message in the sequence.

Reply:
Thanks for the comment!

`findLastSeen` does the search based on sender generated timestamp and this is only needed once (at least for now it is a one time invocation), thus a linear search would be cheaper than sorting all the messages based on their sender generated `timestamp` (messages Index is based on received generated timestamp)

However, we indeed sort messages based on their Index (which encapsulated receiver generated timestamp but not sender generated) redundantly for every incoming history query. We can certainly optimize on that part and sort them once when messages are loaded from db. Will do it in the subsequent PRs.

Reply:
Ah, thanks for clarifying.

Reply:
You are right and the ultimate goal is to use sender generated timestamp, however, that requires a separate PR as it would also affect the pagination logic. It is on the plan though!